### ISSUE TITLE
Invalid function evaluation

### ISSUE THREAD

**Issue by virgil-serbanuta** [runtimeverification/haskell-backend#3738]
The ~Booster~  Haskell backend applies a simplification rule that should not be applied because the `requires` clause is not true. This causes an infinite evaluation loop. ~Actually, it's likely that the Haskell backend does that, but I'm just calling the Booster backend and seeing the results.~

The longer story (reproduction details at the end):

While doing property verification for Elrond's pair contract with the Kasmer tool, the execution reaches a point where it has to evaluate a certain expression (hunting through the equation evaluation logs is time consuming, so I didn't try to identify it). When evaluating this expression, the backend tries to evaluate
```
#getRange(
    SBChunk(#empty(expression-1)) SBChunk(#bytes(b"\x1c\x00\x00\x00")) SBChunk(#bytes(b"\x01")),
    expression-1,
    4
)
```
To do that, it tries to apply this rule ( https://github.com/runtimeverification/mx-backend/blob/pair-lemmas/kmxwasm/k-src/mx-lemmas.md?plain=1#L122-L125 ):
```
  rule #getRange(SBChunk(A:SBItem) B:SparseBytes, Start, Width)
      => #splitGetRange(SBChunk(A) B, Start, size(SBChunk(A)) -Int Start, Start +Int Width -Int size(SBChunk(A)))
    requires Start <Int size(SBChunk(A)) andBool size(SBChunk(A)) <Int Start +Int Width
```
It should not be able to apply this rule because `SBChunk(#empty(expression-1))` evaluates to `expression-1`, and it is `false` that `expression-1 <Int expression-1`.

However, `expression-1` is actually `expression-2 +Int 9974`, so when evaluating this, the backend tries to apply various simplification rules, including this one (https://github.com/runtimeverification/wasm-semantics/blob/master/kwasm-lemmas.md?plain=1#L289):
```
    rule VAL1 +Int VAL2                     <Int MAX => true requires VAL1 <Int MAX andBool VAL2 <Int MAX andBool #distinctBits(VAL1, VAL2) [simplification]
```

The backend quickly evaluates `expression-2 <Int expression-2 +Int 9974` and `9974 <Int expression-2 +Int 9974` to true. However, this rule should not apply because the `#distinctBits(VAL1, VAL2)` function evaluates to true only in a very specific case (see this: #distinctBits(VAL1, VAL2) ). Indeed, the logs show that the backend attempts to evaluate `#distinctBits(expression-2 +Int 9974, expression-2 +Int 9974)` and fails, because the simplification rule does not apply.

Regardless, the backend happily applies the rule above and evaluates `expression-1 <Int expression-1` to true. Then it applies the `#getRange` rule above and produces `#splitGetRange(bytes-stuff, expression-1, 0, 4)` whose evaluation loops because the third arg should never be `0`.

Versions:
```
$ kompile --version
K version:    v6.1.82-0-g8c99ee7740
Build date:   Jo feb 22 23:25:11 EET 2024

$ kore-rpc-booster --version
hs-backend-booster version:
  revision:     6067fee77da53ac357ce5d17ad34e4639070804a
  branch:       HEAD
  last commit:  Tue Jan 16 09:27:16 2024 -0700
```

Bug report:
[weird-function-eval.tar.gz](https://github.com/runtimeverification/hs-backend-booster/files/14380439/weird-function-eval.tar.gz)

**Comment by virgil-serbanuta** [comment]
Yes, you're right, when I said `stuff - stuff` I actually meant that the two things eventually evaluate to the same thing, and the third argument becomes 0.

**Comment by jberthold** [comment]
Unfortunately I wasn't able to craft a smaller example (a simplify request, actually) which would expose the same supposedly-invalid simplification.
I inspected the log a bit more and transformed it into a more readable format that shows the nested simplifications/evaluations - [attaching this file here: looping-server.evaluationtree.txt.gz].(https://github.com/runtimeverification/haskell-backend/files/14518119/looping-server.evaluationtree.txt.gz)
Maybe you can find something that is not right in this view of the evaluation? Ideally we would be able to isolate a term that exposes the behjaviour in a _simplify_ request.

The root cause of this will most likely be a problem with the `distinctBits` predicate. It is abstracted over when the conditions go to the SMT solver (this is visible in the SMT solver transcript). I suspect that this results in a condition check returning a  positive result where it should remain undetermined.

**Comment by virgil-serbanuta** [comment]
Is there a way to get the side condition logged with each of the `DebugAttemptEquation` and `DebugApplyEquation` log entries? Right now I can only get that indirectly, whenever z3 is being invoked. I noticed that it changes while evaluating the term (which could be fine, I guess), but, although I tried, I could not reproduce the exact changes that happen here in a simpler semantics and term. I could try again, but it's harder without being able to see what happens at each step.

Some code that I would need to patch in a Haskell backend and recompile would work for me.

**Comment by jberthold** [comment]
OK, I made a branch where the equation application attempt also logs the side conditions.
You can work with [branch `jb/tracing-additions-for-debugging-side-condition-smt`](https://github.com/runtimeverification/haskell-backend/compare/jb/tracing-additions-for-debugging-side-condition-smt) (build requires GHC-9.6.4 and a recent `stack`, or you can use `nix`)
This is a branch off an earlier master before changing the `overload` attribute name (your `definition.kore` required this earlier state).

Unfortunately, I _still_ don't find any trace of the `distinctBits` function [in the resulting log file: looping-server-with-conditions.evaluationtree.txt.gz](https://github.com/runtimeverification/haskell-backend/files/14531320/looping-server-with-conditions.evaluationtree.txt.gz) - it does not exist but the backend happily applies the rule `wasm-semantics/kwasm-lemmas.md:289:10`  which uses it in the side condition.

**Comment by virgil-serbanuta** [comment]
Thanks, the logging helped, I managed to find this example:

a.k:
```k
module A
  imports BOOL
  imports BYTES
  imports INT

  syntax A ::= a(Int) | b(Int)
  rule a(I) => b(I) requires #bool(f(I)) <=Int 0
  rule a(I) => b(I) requires #bool(f(I)) >Int 0

  syntax Bool ::= f(Int)  [function, total]
  rule f(0) => true
  rule f(I) => loop(I)  [owise]

  syntax Bool ::= g(Int)  [function]
  rule g(I) => true requires I <Int 0  [simplification]

  rule f(I) => f(I) requires g(I)  [simplification]

  syntax Bool ::= loop(Int)  [function]
  syntax Bool ::= loop1(Int)  [function]
  rule loop(I) => loop1(I) requires g(I)  [simplification]
  rule loop1(I) => loop(I) requires g(I)  [simplification]

  syntax Int ::= #bool(Bool)  [function, total]
  rule #bool(B) => 0 requires notBool B
  rule #bool(B) => 1 requires B

endmodule
```

spec.k:
```k
module SPEC
  imports A

  claim <k> a(#bool(f(I))) => b(#bool(f(I))) ... </k>
endmodule
```

command line:
```sh
kompile a.k --backend haskell && kprove spec.k
```

**Comment by geo2a** [comment]
Thanks again @virgil-serbanuta for the minimized example! Looking at it has been deeply insightful, and revealed the bug which is fixed by #3745 (provided it does not break anything else).

The problem is that we're calling the simplifier assuming the remainder predicate, with the hope that the configuration will simplify to `#Bottom` if the remainder configuration is unreachable. This is sound, but overly optimistic, since not every `#Bottom` terminates --- and this is what we are observing here.

The fix is to check that the rule remainder condition is satisfiable, and only then jump into simpifying the configuration.
